src/java/org/apache/cassandra/index/sai/disk/v3/UpperPostings.java

```diff
@@ -0,0 +1,500 @@
+            final Meta meta = Meta.create(compMeta.attributes);
+
+            this.baseUnit = meta.baseUnit;
+
+            for (Map.Entry<Integer,NumericValuesMeta> entry : meta.levelMetas.entrySet())
+            {
+                final MonotonicBlockPackedReader numericReader = new MonotonicBlockPackedReader(upperPostingsOffsetsFile, entry.getValue());
+                final LongArray array = numericReader.open();
+                levelFPMap.put(entry.getKey(), array);
+            }
+
+            this.upperPostingsFile = reader.indexFiles.getFileAndCache(IndexComponent.BLOCK_UPPER_POSTINGS);
+        }
+
+        @Override
+        public void close() throws Exception
+        {
+            for (LongArray array : levelFPMap.values())
+                array.close();
+            upperPostingsFile.close();
+        }
+
+        @SuppressWarnings("resource")
+        public void search(final long startPoint,
+                           final long endPoint,
+                           final PriorityQueue<PostingList.PeekablePostingList> queue) throws IOException
+        {
+            final IndexInput upperPostingsInput = IndexInputReader.create(upperPostingsFile);
+            final IndexInput postingsInput = IndexInputReader.create(reader.postingsHandle);
+
+            final int firstLevel = 0;
+            final int lastLevel = levelFPMap.lastKey();
+
+            long gatheredStart = Integer.MAX_VALUE;
+            long gatheredEnd = -1;
+
+            int postingsCount = 0;
+
+            for (int level = lastLevel; level >= firstLevel; level--)
+            {
+                final LongArray levelPostingFPs = levelFPMap.get(level);
+
+                final int levelUnit = (int) Math.pow(baseUnit, level);
+                final long levelLength = levelUnit * (long)reader.meta.postingsBlockSize;
+
+                final int startDerivedIdx = (int)(startPoint / levelLength);
+                final int endDerivedIdx = (int)(endPoint / levelLength);
+
+                final long levelGatheredStart = gatheredStart;
+                final long levelGatheredEnd = gatheredEnd;
+
+                boolean resetToEnd = false;
+                // iterate forwards over the level file pointers
+                // process the first section
+                // then skip to the last section
+                for (int x = startDerivedIdx; x < levelPostingFPs.length(); x++)
+                {
+                    if (x > endDerivedIdx)
+                        break;
+
+                    final long startPoint2 = x * levelLength;
+                    final long endPoint2 = (x * levelLength) + levelLength - 1;
+
+                    if ((startPoint2 >= startPoint && endPoint2 <= endPoint)
+                        && (startPoint2 < levelGatheredStart || endPoint2 > levelGatheredEnd))
+                    {
+                        System.out.println("level=" + level
```

Comment:
Make sure those printlns are replaced by logging or removed in the final version